# Підготовка даних

[Оригінальний блокнот із *Data Science: Introduction to Machine Learning for Data Science Python and Machine Learning Studio by Lee Stott*](https://github.com/leestott/intro-Datascience/blob/master/Course%20Materials/4-Cleaning_and_Manipulating-Reference.ipynb)

## Дослідження інформації про `DataFrame`

> **Мета навчання:** До кінця цього підрозділу ви повинні впевнено знаходити загальну інформацію про дані, збережені в pandas DataFrames.

Коли ви завантажуєте свої дані в pandas, вони, швидше за все, будуть у форматі `DataFrame`. Але якщо набір даних у вашому `DataFrame` містить 60,000 рядків і 400 стовпців, як взагалі почати розуміти, з чим ви працюєте? На щастя, pandas надає зручні інструменти для швидкого перегляду загальної інформації про `DataFrame`, а також перших і останніх кількох рядків.

Щоб дослідити цю функціональність, ми імпортуємо бібліотеку Python scikit-learn і використаємо знаковий набір даних, який кожен дата-сайєнтист бачив сотні разів: набір даних британського біолога Рональда Фішера *Iris*, використаний у його статті 1936 року "Використання множинних вимірювань у таксономічних проблемах":


In [1]:
import pandas as pd
from sklearn.datasets import load_iris

iris = load_iris()
iris_df = pd.DataFrame(data=iris['data'], columns=iris['feature_names'])

### `DataFrame.shape`
Ми завантажили набір даних Iris у змінну `iris_df`. Перед тим як заглиблюватися в дані, було б корисно дізнатися кількість точок даних, які ми маємо, і загальний розмір набору даних. Це допоможе оцінити обсяг даних, з якими ми працюємо.


In [2]:
iris_df.shape

(150, 4)

Отже, ми маємо справу з 150 рядками та 4 стовпцями даних. Кожен рядок представляє одну точку даних, а кожен стовпець представляє одну ознаку, пов'язану з датафреймом. Тобто, є 150 точок даних, кожна з яких містить 4 ознаки.

`shape` тут є атрибутом датафрейму, а не функцією, тому він не закінчується парою круглих дужок.


### `DataFrame.columns`
Тепер перейдемо до 4 колонок даних. Що саме кожна з них представляє? Атрибут `columns` надасть нам назви колонок у датафреймі.


In [3]:
iris_df.columns

Index(['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)',
       'petal width (cm)'],
      dtype='object')

Як ми бачимо, є чотири (4) стовпці. Атрибут `columns` повідомляє нам назви стовпців і, в основному, нічого більше. Цей атрибут набуває важливості, коли ми хочемо визначити характеристики, які містить набір даних.


### `DataFrame.info`
Кількість даних (визначається атрибутом `shape`) і назви ознак або стовпців (визначаються атрибутом `columns`) дають нам певну інформацію про набір даних. Тепер ми хотіли б заглибитися в набір даних. Функція `DataFrame.info()` є досить корисною для цього.


In [4]:
iris_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 4 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   sepal length (cm)  150 non-null    float64
 1   sepal width (cm)   150 non-null    float64
 2   petal length (cm)  150 non-null    float64
 3   petal width (cm)   150 non-null    float64
dtypes: float64(4)
memory usage: 4.8 KB


Звідси ми можемо зробити кілька спостережень:
1. Тип даних кожного стовпця: У цьому наборі даних вся інформація зберігається у вигляді 64-бітних чисел з плаваючою точкою.
2. Кількість ненульових значень: Робота з нульовими значеннями є важливим етапом підготовки даних. Це буде розглянуто пізніше в блокноті.


### DataFrame.describe()
Припустимо, у нас є багато числових даних у нашому наборі даних. Одновимірні статистичні розрахунки, такі як середнє, медіана, квартилі тощо, можуть бути виконані для кожного стовпця окремо. Функція `DataFrame.describe()` надає нам статистичний огляд числових стовпців набору даних.


In [5]:
iris_df.describe()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
count         150.000000        150.000000         150.000000        150.000000
mean            5.843333          3.057333           3.758000          1.199333
std             0.828066          0.435866           1.765298          0.762238
min             4.300000          2.000000           1.000000          0.100000
25%             5.100000          2.800000           1.600000          0.300000
50%             5.800000          3.000000           4.350000          1.300000
75%             6.400000          3.300000           5.100000          1.800000
max             7.900000          4.400000           6.900000          2.500000

Вихідні дані вище показують загальну кількість точок даних, середнє значення, стандартне відхилення, мінімальне значення, нижній квартиль (25%), медіану (50%), верхній квартиль (75%) та максимальне значення кожного стовпця.


### `DataFrame.head`
З усіма наведеними функціями та атрибутами ми отримали загальний огляд набору даних. Ми знаємо, скільки точок даних є, скільки ознак, тип даних кожної ознаки та кількість ненульових значень для кожної ознаки.

Тепер настав час поглянути на самі дані. Давайте подивимося, як виглядають перші кілька рядків (перші кілька точок даних) нашого `DataFrame`:


In [6]:
iris_df.head()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                5.1               3.5                1.4               0.2
1                4.9               3.0                1.4               0.2
2                4.7               3.2                1.3               0.2
3                4.6               3.1                1.5               0.2
4                5.0               3.6                1.4               0.2

Як видно з результату, тут є п'ять (5) записів набору даних. Якщо подивитися на індекс зліва, ми дізнаємося, що це перші п'ять рядків.


### Вправа:

З наведеного вище прикладу зрозуміло, що за замовчуванням `DataFrame.head` повертає перші п'ять рядків `DataFrame`. У кодовій комірці нижче, чи можете ви знайти спосіб відобразити більше ніж п'ять рядків?


In [7]:
# Hint: Consult the documentation by using iris_df.head?

### `DataFrame.tail`
Інший спосіб перегляду даних — це почати з кінця (замість початку). Протилежністю `DataFrame.head` є `DataFrame.tail`, який повертає останні п'ять рядків `DataFrame`:


In [8]:
iris_df.tail()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
145                6.7               3.0                5.2               2.3
146                6.3               2.5                5.0               1.9
147                6.5               3.0                5.2               2.0
148                6.2               3.4                5.4               2.3
149                5.9               3.0                5.1               1.8

На практиці корисно мати можливість легко переглядати перші кілька рядків або останні кілька рядків `DataFrame`, особливо коли ви шукаєте аномалії в упорядкованих наборах даних.

Усі функції та атрибути, показані вище за допомогою прикладів коду, допомагають нам отримати уявлення про дані.

> **Висновок:** Навіть просто переглянувши метадані про інформацію в DataFrame або перші й останні кілька значень, ви можете одразу отримати уявлення про розмір, форму та зміст даних, з якими працюєте.


### Відсутні дані
Давайте розглянемо тему відсутніх даних. Відсутні дані виникають, коли в деяких стовпцях не збережено жодного значення.

Розглянемо приклад: припустимо, хтось дуже переймається своєю вагою і не заповнює поле "вага" в опитуванні. У такому випадку значення ваги для цієї людини буде відсутнім.

У більшості випадків у реальних наборах даних трапляються відсутні значення.

**Як Pandas обробляє відсутні дані**

Pandas обробляє відсутні значення двома способами. Перший спосіб ви вже бачили в попередніх розділах: `NaN`, або Not a Number. Це насправді спеціальне значення, яке є частиною специфікації IEEE для чисел з плаваючою точкою, і воно використовується лише для позначення відсутніх значень типу float.

Для відсутніх значень, які не є числами з плаваючою точкою, pandas використовує об'єкт Python `None`. Хоча може здатися заплутаним, що ви зустрічаєте два різні типи значень, які фактично означають одне й те саме, існують обґрунтовані програмні причини для такого вибору дизайну. На практиці це дозволяє pandas забезпечити хороший компроміс для переважної більшості випадків. Незважаючи на це, як `None`, так і `NaN` мають обмеження, про які потрібно пам'ятати щодо того, як їх можна використовувати.


### `None`: відсутні дані, які не є числами з плаваючою точкою
Оскільки `None` походить з Python, його не можна використовувати в масивах NumPy і pandas, які не мають типу даних `'object'`. Пам’ятайте, що масиви NumPy (і структури даних у pandas) можуть містити лише один тип даних. Це надає їм величезну потужність для роботи з великими обсягами даних і обчислень, але також обмежує їхню гнучкість. Такі масиви повинні бути приведені до "найнижчого спільного знаменника", тобто типу даних, який охоплює все в масиві. Коли в масиві є `None`, це означає, що ви працюєте з об'єктами Python.

Щоб побачити це на практиці, розгляньте наступний приклад масиву (зверніть увагу на його `dtype`):


In [9]:
import numpy as np

example1 = np.array([2, None, 6, 8])
example1

array([2, None, 6, 8], dtype=object)

Реальність підвищення типів даних має два побічні ефекти. По-перше, операції виконуватимуться на рівні інтерпретованого коду Python, а не компільованого коду NumPy. По суті, це означає, що будь-які операції, які включають `Series` або `DataFrames` із `None`, будуть повільнішими. Хоча ви, ймовірно, не помітите цього впливу на продуктивність, для великих наборів даних це може стати проблемою.

Другий побічний ефект випливає з першого. Оскільки `None` фактично повертає `Series` або `DataFrame` у світ звичайного Python, використання агрегатів NumPy/pandas, таких як `sum()` або `min()` на масивах, які містять значення ``None``, зазвичай призведе до помилки:


In [10]:
example1.sum()

TypeError: ignored

**Основний висновок**: Додавання (та інші операції) між цілими числами та значеннями `None` є невизначеним, що може обмежити можливості роботи з наборами даних, які їх містять.


### `NaN`: відсутні значення типу float

На відміну від `None`, NumPy (а отже і pandas) підтримує `NaN` для швидких, векторизованих операцій та ufuncs. Погана новина полягає в тому, що будь-яка арифметична операція з `NaN` завжди дає результат `NaN`. Наприклад:


In [11]:
np.nan + 1

nan

In [12]:
np.nan * 0

nan

Хороша новина: агрегації, які виконуються на масивах з `NaN`, не викликають помилок. Погана новина: результати не завжди корисні:


In [13]:
example2 = np.array([2, np.nan, 6, 8]) 
example2.sum(), example2.min(), example2.max()

(nan, nan, nan)

### Вправа:


In [11]:
# What happens if you add np.nan and None together?


Пам'ятайте: `NaN` призначений лише для відсутніх значень з плаваючою точкою; немає еквівалента `NaN` для цілих чисел, рядків або булевих значень.


### `NaN` і `None`: нульові значення в pandas

Хоча `NaN` і `None` можуть поводитися трохи по-різному, pandas все ж створений для роботи з ними як взаємозамінними. Щоб зрозуміти, що мається на увазі, розглянемо `Series` цілих чисел:


In [15]:
int_series = pd.Series([1, 2, 3], dtype=int)
int_series

0    1
1    2
2    3
dtype: int64

### Вправа:


In [16]:
# Now set an element of int_series equal to None.
# How does that element show up in the Series?
# What is the dtype of the Series?


У процесі приведення типів даних до однорідності в `Series` та `DataFrame`, pandas легко змінює відсутні значення між `None` та `NaN`. Через цю особливість дизайну корисно розглядати `None` та `NaN` як два різні види "null" у pandas. Насправді, деякі основні методи, які ви будете використовувати для роботи з відсутніми значеннями в pandas, відображають цю ідею у своїх назвах:

- `isnull()`: Створює булеву маску, яка вказує на відсутні значення
- `notnull()`: Протилежність `isnull()`
- `dropna()`: Повертає відфільтровану версію даних
- `fillna()`: Повертає копію даних із заповненими або заміненими відсутніми значеннями

Це важливі методи, які варто освоїти та використовувати впевнено, тому давайте розглянемо кожен із них детальніше.


### Виявлення нульових значень

Тепер, коли ми зрозуміли важливість пропущених значень, нам потрібно виявити їх у нашому наборі даних, перш ніж працювати з ними. 
Методи `isnull()` і `notnull()` є основними для виявлення нульових даних. Обидва повертають булеві маски для ваших даних.


In [17]:
example3 = pd.Series([0, np.nan, '', None])

In [18]:
example3.isnull()

0    False
1     True
2    False
3     True
dtype: bool

Уважно подивіться на результат. Чи щось у ньому вас здивувало? Хоча `0` є арифметичним нулем, він все ж таки є цілком коректним цілим числом, і pandas сприймає його саме так. `''` є трохи складнішим випадком. Хоча ми використовували його в Розділі 1 для позначення порожнього рядка, він все ж таки є об'єктом типу рядок, а не представленням null з точки зору pandas.

Тепер давайте розглянемо це з іншого боку і використаємо ці методи так, як ви, ймовірно, будете використовувати їх на практиці. Ви можете використовувати булеві маски безпосередньо як індекс для ``Series`` або ``DataFrame``, що може бути корисним, коли потрібно працювати з ізольованими відсутніми (або наявними) значеннями.

Якщо ми хочемо отримати загальну кількість відсутніх значень, ми можемо просто виконати сумування за маскою, створеною методом `isnull()`.


In [19]:
example3.isnull().sum()

2

### Вправа:


In [20]:
# Try running example3[example3.notnull()].
# Before you do so, what do you expect to see?


**Основний висновок**: Обидва методи `isnull()` та `notnull()` дають схожі результати при використанні в DataFrame: вони показують результати та індекс цих результатів, що значно допоможе вам у роботі з вашими даними.


### Робота з відсутніми даними

> **Мета навчання:** Після завершення цього підрозділу ви повинні знати, як і коли замінювати або видаляти нульові значення з DataFrames.

Моделі машинного навчання не можуть самостійно працювати з відсутніми даними. Тому перед тим, як передати дані в модель, необхідно вирішити проблему з цими відсутніми значеннями.

Те, як обробляються відсутні дані, має тонкі компроміси, може вплинути на ваш остаточний аналіз і результати в реальному світі.

Існує два основних способи вирішення проблеми з відсутніми даними:

1.   Видалити рядок, що містить відсутнє значення
2.   Замінити відсутнє значення на інше значення

Ми обговоримо обидва ці методи, а також їхні переваги та недоліки детально.


### Видалення нульових значень

Кількість даних, які ми передаємо нашій моделі, безпосередньо впливає на її продуктивність. Видалення нульових значень означає, що ми зменшуємо кількість точок даних, а отже, зменшуємо розмір набору даних. Тому рекомендується видаляти рядки з нульовими значеннями, якщо набір даних досить великий.

Інший випадок може бути, коли певний рядок або стовпець має багато відсутніх значень. Тоді їх можна видалити, оскільки вони не додадуть великої цінності до нашого аналізу, адже більшість даних для цього рядка/стовпця відсутні.

Окрім виявлення відсутніх значень, pandas надає зручний спосіб видалення нульових значень із `Series` та `DataFrame`. Щоб побачити це на практиці, повернемося до `example3`. Функція `DataFrame.dropna()` допомагає видаляти рядки з нульовими значеннями.


In [21]:
example3 = example3.dropna()
example3

0    0
2     
dtype: object

Зверніть увагу, що це має виглядати як ваш результат з `example3[example3.notnull()]`. Різниця тут полягає в тому, що, замість просто індексації за маскованими значеннями, `dropna` видалив ці відсутні значення з `Series` `example3`.

Оскільки DataFrame має два виміри, він надає більше варіантів для видалення даних.


In [22]:
example4 = pd.DataFrame([[1,      np.nan, 7], 
                         [2,      5,      8], 
                         [np.nan, 6,      9]])
example4

0    1  2
0  1.0  NaN  7
1  2.0  5.0  8
2  NaN  6.0  9

(Чи помітили ви, що pandas перетворив два стовпці на тип float, щоб врахувати `NaN`?)

Ви не можете видалити окреме значення з `DataFrame`, тому доведеться видаляти цілі рядки або стовпці. Залежно від того, що ви робите, вам може знадобитися одне або інше, і pandas надає вам можливості для обох варіантів. Оскільки в аналізі даних стовпці зазвичай представляють змінні, а рядки — спостереження, частіше за все видаляють рядки даних; налаштування за замовчуванням для `dropna()` — видаляти всі рядки, які містять будь-які null-значення:


In [23]:
example4.dropna()

0    1  2
1  2.0  5.0  8

Якщо необхідно, ви можете видалити значення NA з колонок. Використовуйте `axis=1`, щоб зробити це:


In [24]:
example4.dropna(axis='columns')

2
0  7
1  8
2  9

Зверніть увагу, що це може призвести до втрати великої кількості даних, які ви, можливо, хочете зберегти, особливо у менших наборах даних. Що робити, якщо ви хочете видалити лише ті рядки або стовпці, які містять кілька або навіть усі значення, що дорівнюють null? Ви можете налаштувати ці параметри у `dropna` за допомогою параметрів `how` і `thresh`.

За замовчуванням `how='any'` (якщо ви хочете перевірити самостійно або побачити, які інші параметри має цей метод, виконайте `example4.dropna?` у кодовій комірці). Ви також можете вказати `how='all'`, щоб видалити лише ті рядки або стовпці, які містять усі значення null. Давайте розширимо наш приклад `DataFrame`, щоб побачити це в дії у наступній вправі.


In [25]:
example4[3] = np.nan
example4

0    1  2   3
0  1.0  NaN  7 NaN
1  2.0  5.0  8 NaN
2  NaN  6.0  9 NaN

> Основні моменти:
1. Видаляти порожні значення варто лише тоді, коли набір даних достатньо великий.
2. Повні рядки або стовпці можна видаляти, якщо більшість їхніх даних відсутні.
3. Метод `DataFrame.dropna(axis=)` допомагає видаляти порожні значення. Аргумент `axis` визначає, чи потрібно видаляти рядки, чи стовпці.
4. Можна також використовувати аргумент `how`. За замовчуванням він встановлений на `any`, тому видаляються лише ті рядки/стовпці, які містять будь-які порожні значення. Його можна встановити на `all`, щоб вказати, що будуть видалені лише ті рядки/стовпці, де всі значення порожні.


### Вправа:


In [22]:
# How might you go about dropping just column 3?
# Hint: remember that you will need to supply both the axis parameter and the how parameter.


Параметр `thresh` надає вам більш детальний контроль: ви встановлюєте кількість *не-null* значень, які рядок або стовпець повинні мати, щоб бути збереженими:


In [27]:
example4.dropna(axis='rows', thresh=3)

0    1  2   3
1  2.0  5.0  8 NaN

Тут перший і останній рядки були видалені, оскільки вони містять лише два ненульові значення.


### Заповнення пропущених значень

Іноді має сенс заповнити пропущені значення такими, які можуть бути дійсними. Існує кілька методів для заповнення null-значень. Перший — використання галузевих знань (знання предметної області, на якій базується набір даних) для приблизного визначення пропущених значень.

Ви можете використовувати `isnull` для заповнення значень безпосередньо, але це може бути трудомістким, особливо якщо потрібно заповнити багато значень. Оскільки це дуже поширене завдання в аналізі даних, pandas пропонує метод `fillna`, який повертає копію `Series` або `DataFrame` з пропущеними значеннями, заміненими на ті, які ви обрали. Давайте створимо ще один приклад `Series`, щоб побачити, як це працює на практиці.


### Категоричні дані (нечислові)
Спочатку розглянемо нечислові дані. У наборах даних є стовпці з категоричними даними, наприклад, стать, істина чи хибність тощо.

У більшості таких випадків ми замінюємо пропущені значення на `моду` стовпця. Наприклад, у нас є 100 точок даних, з яких 90 відповіли "Істина", 8 — "Хибність", а 2 не заповнили. Тоді ми можемо заповнити ці 2 значення "Істиною", враховуючи весь стовпець.

Знову ж таки, тут можна використовувати знання предметної області. Розглянемо приклад заповнення за модою.


In [28]:
fill_with_mode = pd.DataFrame([[1,2,"True"],
                               [3,4,None],
                               [5,6,"False"],
                               [7,8,"True"],
                               [9,10,"True"]])

fill_with_mode

0   1      2
0  1   2   True
1  3   4   None
2  5   6  False
3  7   8   True
4  9  10   True

Тепер спочатку знайдемо моду перед заповненням значення `None` модою.


In [29]:
fill_with_mode[2].value_counts()

True     3
False    1
Name: 2, dtype: int64

Отже, ми замінимо None на True


In [30]:
fill_with_mode[2].fillna('True',inplace=True)

In [31]:
fill_with_mode

0   1      2
0  1   2   True
1  3   4   True
2  5   6  False
3  7   8   True
4  9  10   True

Як ми бачимо, значення null було замінено. Безумовно, ми могли написати будь-що замість `'True'`, і це було б підставлено.


### Числові дані
Тепер перейдемо до числових даних. Тут є два поширені способи заміни відсутніх значень:

1. Замінити на медіану рядка
2. Замінити на середнє значення рядка

Медіану використовують у випадку, якщо дані мають перекіс і містять викиди. Це тому, що медіана стійка до впливу викидів.

Коли дані нормалізовані, можна використовувати середнє значення, оскільки в такому випадку середнє і медіана будуть досить близькими.

Спочатку візьмемо стовпець, який має нормальний розподіл, і заповнимо відсутнє значення середнім значенням цього стовпця.


In [32]:
fill_with_mean = pd.DataFrame([[-2,0,1],
                               [-1,2,3],
                               [np.nan,4,5],
                               [1,6,7],
                               [2,8,9]])

fill_with_mean

0  1  2
0 -2.0  0  1
1 -1.0  2  3
2  NaN  4  5
3  1.0  6  7
4  2.0  8  9

Середнє значення стовпця дорівнює


In [33]:
np.mean(fill_with_mean[0])

0.0

Заповнення середнім значенням


In [34]:
fill_with_mean[0].fillna(np.mean(fill_with_mean[0]),inplace=True)
fill_with_mean

0  1  2
0 -2.0  0  1
1 -1.0  2  3
2  0.0  4  5
3  1.0  6  7
4  2.0  8  9

Як ми бачимо, відсутнє значення було замінено його середнім значенням.


Тепер давайте спробуємо інший датафрейм, і цього разу ми замінимо значення None на медіану стовпця.


In [35]:
fill_with_median = pd.DataFrame([[-2,0,1],
                               [-1,2,3],
                               [0,np.nan,5],
                               [1,6,7],
                               [2,8,9]])

fill_with_median

0    1  2
0 -2  0.0  1
1 -1  2.0  3
2  0  NaN  5
3  1  6.0  7
4  2  8.0  9

Медіана другого стовпця становить


In [36]:
fill_with_median[1].median()

4.0

Заповнення медіаною


In [37]:
fill_with_median[1].fillna(fill_with_median[1].median(),inplace=True)
fill_with_median

0    1  2
0 -2  0.0  1
1 -1  2.0  3
2  0  4.0  5
3  1  6.0  7
4  2  8.0  9

Як ми можемо бачити, значення NaN було замінено на медіану стовпця


In [38]:
example5 = pd.Series([1, np.nan, 2, None, 3], index=list('abcde'))
example5

a    1.0
b    NaN
c    2.0
d    NaN
e    3.0
dtype: float64

Ви можете заповнити всі порожні записи одним значенням, наприклад, `0`:


In [39]:
example5.fillna(0)

a    1.0
b    0.0
c    2.0
d    0.0
e    3.0
dtype: float64

> Основні висновки:
1. Заповнення відсутніх значень слід виконувати, коли даних недостатньо або є стратегія для заповнення пропусків.
2. Домашні знання можуть бути використані для заповнення відсутніх значень шляхом їх наближення.
3. Для категоріальних даних зазвичай відсутні значення замінюються модою стовпця.
4. Для числових даних відсутні значення зазвичай заповнюються середнім значенням (для нормалізованих наборів даних) або медіаною стовпців.


### Вправа:


In [40]:
# What happens if you try to fill null values with a string, like ''?


Ви можете **заповнити вперед** null-значення, використовуючи останнє допустиме значення для заповнення null:


In [41]:
example5.fillna(method='ffill')

a    1.0
b    1.0
c    2.0
d    2.0
e    3.0
dtype: float64

Ви також можете **заповнити назад**, щоб поширити наступне допустиме значення назад для заповнення null:


In [42]:
example5.fillna(method='bfill')

a    1.0
b    2.0
c    2.0
d    3.0
e    3.0
dtype: float64

Як ви могли здогадатися, це працює так само з DataFrames, але ви також можете вказати `axis`, уздовж якого заповнювати нульові значення:


In [43]:
example4

0    1  2   3
0  1.0  NaN  7 NaN
1  2.0  5.0  8 NaN
2  NaN  6.0  9 NaN

In [44]:
example4.fillna(method='ffill', axis=1)

0    1    2    3
0  1.0  1.0  7.0  7.0
1  2.0  5.0  8.0  8.0
2  NaN  6.0  9.0  9.0

Зверніть увагу, що коли попереднє значення недоступне для заповнення вперед, нульове значення залишається.


### Вправа:


In [45]:
# What output does example4.fillna(method='bfill', axis=1) produce?
# What about example4.fillna(method='ffill') or example4.fillna(method='bfill')?
# Can you think of a longer code snippet to write that can fill all of the null values in example4?


Ви можете творчо використовувати `fillna`. Наприклад, давайте знову розглянемо `example4`, але цього разу заповнимо відсутні значення середнім значенням усіх значень у `DataFrame`:


In [46]:
example4.fillna(example4.mean())

0    1  2   3
0  1.0  5.5  7 NaN
1  2.0  5.0  8 NaN
2  1.5  6.0  9 NaN

Зверніть увагу, що стовпець 3 все ще не має значень: за замовчуванням значення заповнюються по рядках.

> **Висновок:** Існує кілька способів вирішення проблеми з відсутніми значеннями у ваших наборах даних. Конкретна стратегія (видалення, заміна або спосіб заміни) повинна залежати від особливостей цих даних. Ви краще зрозумієте, як працювати з відсутніми значеннями, чим більше будете взаємодіяти з наборами даних.


### Кодування категоріальних даних

Моделі машинного навчання працюють лише з числами та будь-якою формою числових даних. Вони не зможуть розрізнити "Так" і "Ні", але зможуть відрізнити 0 від 1. Тому після заповнення пропущених значень необхідно закодувати категоріальні дані в числову форму, щоб модель могла їх зрозуміти.

Кодування можна виконати двома способами. Ми розглянемо їх далі.


**КОДУВАННЯ МІТКИ**

Кодування мітки — це процес перетворення кожної категорії в число. Наприклад, уявімо, що у нас є набір даних про пасажирів авіаліній, і є стовпець, який містить їхній клас серед наступних ['бізнес-клас', 'економ-клас', 'перший клас']. Якщо виконати кодування мітки для цього, то це буде перетворено на [0,1,2]. Давайте розглянемо приклад за допомогою коду. Оскільки ми будемо вивчати `scikit-learn` у наступних блокнотах, тут ми його використовувати не будемо.


In [47]:
label = pd.DataFrame([
                      [10,'business class'],
                      [20,'first class'],
                      [30, 'economy class'],
                      [40, 'economy class'],
                      [50, 'economy class'],
                      [60, 'business class']
],columns=['ID','class'])
label

ID           class
0  10  business class
1  20     first class
2  30   economy class
3  40   economy class
4  50   economy class
5  60  business class

Щоб виконати кодування міток для першого стовпця, спочатку потрібно описати відображення кожного класу на число, перед заміною.


In [48]:
class_labels = {'business class':0,'economy class':1,'first class':2}
label['class'] = label['class'].replace(class_labels)
label

ID  class
0  10      0
1  20      2
2  30      1
3  40      1
4  50      1
5  60      0

Як ми бачимо, результат відповідає тому, що ми очікували. Отже, коли слід використовувати кодування міток? Кодування міток використовується в одному або обох наступних випадках:
1. Коли кількість категорій велика
2. Коли категорії мають порядок.


**КОДУВАННЯ ONE HOT**

Ще одним видом кодування є One Hot Encoding. У цьому типі кодування кожна категорія стовпця додається як окремий стовпець, і кожна точка даних отримує 0 або 1 залежно від того, чи містить вона цю категорію. Отже, якщо є n різних категорій, до датафрейму буде додано n стовпців.

Наприклад, візьмемо той самий приклад класів літака. Категорії були: ['бізнес-клас', 'економ-клас', 'перший клас']. Якщо ми виконаємо кодування One Hot, до набору даних будуть додані наступні три стовпці: ['class_business class', 'class_economy class', 'class_first class'].


In [49]:
one_hot = pd.DataFrame([
                      [10,'business class'],
                      [20,'first class'],
                      [30, 'economy class'],
                      [40, 'economy class'],
                      [50, 'economy class'],
                      [60, 'business class']
],columns=['ID','class'])
one_hot

ID           class
0  10  business class
1  20     first class
2  30   economy class
3  40   economy class
4  50   economy class
5  60  business class

Давайте виконаємо one hot encoding для першого стовпця


In [50]:
one_hot_data = pd.get_dummies(one_hot,columns=['class'])

In [51]:
one_hot_data

ID  class_business class  class_economy class  class_first class
0  10                     1                    0                  0
1  20                     0                    0                  1
2  30                     0                    1                  0
3  40                     0                    1                  0
4  50                     0                    1                  0
5  60                     1                    0                  0

Кожен one-hot закодований стовпець містить 0 або 1, що вказує, чи існує ця категорія для даної точки даних.


Коли ми використовуємо one hot encoding? One hot encoding використовується в одному або обох наступних випадках:

1. Коли кількість категорій і розмір набору даних є невеликими.
2. Коли категорії не мають певного порядку.


> Основні висновки:
1. Кодування здійснюється для перетворення ненумеричних даних у числові.
2. Існує два типи кодування: кодування міток і One Hot кодування, які можна виконувати залежно від вимог набору даних.


## Видалення дубльованих даних

> **Мета навчання:** Після завершення цього підрозділу ви повинні впевнено визначати та видаляти дубльовані значення з DataFrames.

Окрім відсутніх даних, у реальних наборах даних ви часто зустрічатимете дубльовані дані. На щастя, pandas пропонує простий спосіб виявлення та видалення дубльованих записів.


### Виявлення дублікатів: `duplicated`

Ви можете легко знайти дублікати за допомогою методу `duplicated` у pandas, який повертає булеву маску, що вказує, чи є запис у `DataFrame` дублікатом попереднього. Давайте створимо ще один приклад `DataFrame`, щоб побачити це в дії.


In [52]:
example6 = pd.DataFrame({'letters': ['A','B'] * 2 + ['B'],
                         'numbers': [1, 2, 1, 3, 3]})
example6

letters  numbers
0       A        1
1       B        2
2       A        1
3       B        3
4       B        3

In [53]:
example6.duplicated()

0    False
1    False
2     True
3    False
4     True
dtype: bool

### Видалення дублікатів: `drop_duplicates`
`drop_duplicates` просто повертає копію даних, для яких всі значення `duplicated` є `False`:


In [54]:
example6.drop_duplicates()

letters  numbers
0       A        1
1       B        2
3       B        3

І `duplicated`, і `drop_duplicates` за замовчуванням враховують усі стовпці, але ви можете вказати, щоб вони перевіряли лише підмножину стовпців у вашому `DataFrame`:


In [55]:
example6.drop_duplicates(['letters'])

letters  numbers
0       A        1
1       B        2

> **Висновок:** Видалення дубльованих даних є важливою частиною майже кожного проєкту з аналізу даних. Дубльовані дані можуть змінити результати ваших аналізів і надати вам неточні результати!


## Перевірка якості даних у реальному світі

> **Мета навчання:** До кінця цього розділу ви повинні впевнено виявляти та виправляти поширені проблеми якості даних у реальному світі, включаючи непослідовні категоріальні значення, аномальні числові значення (викиди) та дублікати сутностей із варіаціями.

Хоча пропущені значення та точні дублікати є поширеними проблемами, реальні набори даних часто містять більш тонкі недоліки:

1. **Непослідовні категоріальні значення**: Одна й та сама категорія записана по-різному (наприклад, "USA", "U.S.A", "United States")
2. **Аномальні числові значення**: Екстремальні викиди, які вказують на помилки введення даних (наприклад, вік = 999)
3. **Майже дублікати рядків**: Записи, які представляють одну й ту саму сутність із невеликими варіаціями

Давайте розглянемо методи виявлення та усунення цих проблем.


### Створення прикладу "брудного" набору даних

Спочатку створимо приклад набору даних, який містить типові проблеми, з якими ми часто стикаємося в реальних даних:


In [ ]:
import pandas as pd
import numpy as np

# Create a sample dataset with quality issues
dirty_data = pd.DataFrame({
    'customer_id': [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12],
    'name': ['John Smith', 'Jane Doe', 'John Smith', 'Bob Johnson', 
             'Alice Williams', 'Charlie Brown', 'John  Smith', 'Eva Martinez',
             'Bob Johnson', 'Diana Prince', 'Frank Castle', 'Alice Williams'],
    'age': [25, 32, 25, 45, 28, 199, 25, 31, 45, 27, -5, 28],
    'country': ['USA', 'UK', 'U.S.A', 'Canada', 'USA', 'United Kingdom',
                'United States', 'Mexico', 'canada', 'USA', 'UK', 'usa'],
    'purchase_amount': [100.50, 250.00, 105.00, 320.00, 180.00, 90.00,
                       102.00, 275.00, 325.00, 195.00, 410.00, 185.00]
})

print("Sample 'Dirty' Dataset:")
print(dirty_data)

### 1. Виявлення непослідовних категоріальних значень

Зверніть увагу, що в стовпці `country` є кілька варіантів представлення одних і тих самих країн. Давайте визначимо ці непослідовності:


In [ ]:
# Check unique values in the country column
print("Unique country values:")
print(dirty_data['country'].unique())
print(f"\nTotal unique values: {dirty_data['country'].nunique()}")

# Count occurrences of each variation
print("\nValue counts:")
print(dirty_data['country'].value_counts())

#### Стандартизація категоріальних значень

Ми можемо створити відображення для стандартизації цих значень. Простий підхід — перетворити текст у нижній регістр і створити словник відображення:


In [ ]:
# Create a standardization mapping
country_mapping = {
    'usa': 'USA',
    'u.s.a': 'USA',
    'united states': 'USA',
    'uk': 'UK',
    'united kingdom': 'UK',
    'canada': 'Canada',
    'mexico': 'Mexico'
}

# Standardize the country column
dirty_data['country_clean'] = dirty_data['country'].str.lower().map(country_mapping)

print("Before standardization:")
print(dirty_data['country'].value_counts())
print("\nAfter standardization:")
print(dirty_data[['country_clean']].value_counts())

**Альтернатива: Використання нечіткого порівняння**

Для більш складних випадків ми можемо використовувати нечітке порівняння рядків за допомогою бібліотеки `rapidfuzz`, щоб автоматично знаходити схожі рядки:


In [ ]:
try:
    from rapidfuzz import process, fuzz
except ImportError:
    print("rapidfuzz is not installed. Please install it with 'pip install rapidfuzz' to use fuzzy matching.")
    process = None
    fuzz = None

# Get unique countries
unique_countries = dirty_data['country'].unique()

# For each country, find similar matches
if process is not None and fuzz is not None:
    print("Finding similar country names (similarity > 70%):")
    for country in unique_countries:
        matches = process.extract(country, unique_countries, scorer=fuzz.ratio, limit=3)
        # Filter matches with similarity > 70 and not identical
        similar = [m for m in matches if m[1] > 70 and m[0] != country]
        if similar:
            print(f"\n'{country}' is similar to:")
            for match, score, _ in similar:
                print(f"  - '{match}' (similarity: {score}%)")
else:
    print("Skipping fuzzy matching because rapidfuzz is not available.")

### 2. Виявлення аномальних числових значень (викидів)

Розглядаючи стовпець `age`, ми бачимо деякі підозрілі значення, такі як 199 і -5. Давайте використаємо статистичні методи для виявлення цих викидів.


In [ ]:
# Display basic statistics
print("Age column statistics:")
print(dirty_data['age'].describe())

# Identify impossible values using domain knowledge
print("\nRows with impossible age values (< 0 or > 120):")
impossible_ages = dirty_data[(dirty_data['age'] < 0) | (dirty_data['age'] > 120)]
print(impossible_ages[['customer_id', 'name', 'age']])

#### Використання методу IQR (міжквартильний розмах)

Метод IQR — це надійний статистичний підхід для виявлення аномалій, який менш чутливий до екстремальних значень:


In [ ]:
# Calculate IQR for age (excluding impossible values)
valid_ages = dirty_data[(dirty_data['age'] >= 0) & (dirty_data['age'] <= 120)]['age']

Q1 = valid_ages.quantile(0.25)
Q3 = valid_ages.quantile(0.75)
IQR = Q3 - Q1

# Define outlier bounds
lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

print(f"IQR-based outlier bounds for age: [{lower_bound:.2f}, {upper_bound:.2f}]")

# Identify outliers
age_outliers = dirty_data[(dirty_data['age'] < lower_bound) | (dirty_data['age'] > upper_bound)]
print(f"\nRows with age outliers:")
print(age_outliers[['customer_id', 'name', 'age']])

#### Використання методу Z-оцінки

Метод Z-оцінки визначає аномалії на основі стандартних відхилень від середнього значення:


In [ ]:
try:
    from scipy import stats
except ImportError:
    print("scipy is required for Z-score calculation. Please install it with 'pip install scipy' and rerun this cell.")
else:
    # Calculate Z-scores for age, handling NaN values
    age_nonan = dirty_data['age'].dropna()
    zscores = np.abs(stats.zscore(age_nonan))
    dirty_data['age_zscore'] = np.nan
    dirty_data.loc[age_nonan.index, 'age_zscore'] = zscores

    # Typically, Z-score > 3 indicates an outlier
    print("Rows with age Z-score > 3:")
    zscore_outliers = dirty_data[dirty_data['age_zscore'] > 3]
    print(zscore_outliers[['customer_id', 'name', 'age', 'age_zscore']])

    # Clean up the temporary column
    dirty_data = dirty_data.drop('age_zscore', axis=1)

#### Обробка Викидів

Після виявлення викиди можна обробляти кількома способами:
1. **Видалити**: Видалити рядки з викидами (якщо це помилки)
2. **Обрізати**: Замінити на граничні значення
3. **Замінити на NaN**: Розглядати як відсутні дані та використовувати методи імпутації
4. **Залишити**: Якщо це справжні екстремальні значення


In [ ]:
# Create a cleaned version by replacing impossible ages with NaN
dirty_data['age_clean'] = dirty_data['age'].apply(
    lambda x: np.nan if (x < 0 or x > 120) else x
)

print("Age column before and after cleaning:")
print(dirty_data[['customer_id', 'name', 'age', 'age_clean']])

### 3. Виявлення майже дубльованих рядків

Зверніть увагу, що в нашому наборі даних є кілька записів для "John Smith" із трохи різними значеннями. Давайте визначимо потенційні дублікати на основі схожості імен.


In [ ]:
# First, let's look at exact name matches (ignoring extra whitespace)
dirty_data['name_normalized'] = dirty_data['name'].str.strip().str.lower()

print("Checking for duplicate names:")
duplicate_names = dirty_data[dirty_data.duplicated(['name_normalized'], keep=False)]
print(duplicate_names.sort_values('name_normalized')[['customer_id', 'name', 'age', 'country']])

#### Пошук майже дубльованих записів за допомогою нечіткого співставлення

Для більш складного виявлення дубльованих записів можна використовувати нечітке співставлення для пошуку схожих імен:


In [ ]:
try:
    from rapidfuzz import process, fuzz

    # Function to find potential duplicates
    def find_near_duplicates(df, column, threshold=90):
        """
        Find near-duplicate entries in a column using fuzzy matching.
        
        Parameters:
        - df: DataFrame
        - column: Column name to check for duplicates
        - threshold: Similarity threshold (0-100)
        
        Returns: List of potential duplicate groups
        """
        values = df[column].unique()
        duplicate_groups = []
        checked = set()
        
        for value in values:
            if value in checked:
                continue
                
            # Find similar values
            matches = process.extract(value, values, scorer=fuzz.ratio, limit=len(values))
            similar = [m[0] for m in matches if m[1] >= threshold]
            
            if len(similar) > 1:
                duplicate_groups.append(similar)
                checked.update(similar)
        
        return duplicate_groups

    # Find near-duplicate names
    duplicate_groups = find_near_duplicates(dirty_data, 'name', threshold=90)

    print("Potential duplicate groups:")
    for i, group in enumerate(duplicate_groups, 1):
        print(f"\nGroup {i}:")
        for name in group:
            matching_rows = dirty_data[dirty_data['name'] == name]
            print(f"  '{name}': {len(matching_rows)} occurrence(s)")
            for _, row in matching_rows.iterrows():
                print(f"    - Customer {row['customer_id']}: age={row['age']}, country={row['country']}")
except ImportError:
    print("rapidfuzz is not installed. Skipping fuzzy matching for near-duplicates.")

#### Обробка дублікатів

Після виявлення необхідно вирішити, як обробляти дублікати:
1. **Залишити перше входження**: Використовуйте `drop_duplicates(keep='first')`
2. **Залишити останнє входження**: Використовуйте `drop_duplicates(keep='last')`
3. **Агрегувати інформацію**: Об'єднати інформацію з дубльованих рядків
4. **Ручний перегляд**: Позначити для перевірки людиною


In [ ]:
# Example: Remove duplicates based on normalized name, keeping first occurrence
cleaned_data = dirty_data.drop_duplicates(subset=['name_normalized'], keep='first')

print(f"Original dataset: {len(dirty_data)} rows")
print(f"After removing name duplicates: {len(cleaned_data)} rows")
print(f"Removed: {len(dirty_data) - len(cleaned_data)} duplicate rows")

print("\nCleaned dataset:")
print(cleaned_data[['customer_id', 'name', 'age', 'country_clean']])

### Резюме: Повна система очищення даних

Давайте об'єднаємо все в одну комплексну систему очищення даних:


In [ ]:
def clean_dataset(df):
    """
    Comprehensive data cleaning function.
    """
    # Create a copy to avoid modifying the original
    cleaned = df.copy()
    
    # 1. Standardize categorical values (country)
    country_mapping = {
        'usa': 'USA', 'u.s.a': 'USA', 'united states': 'USA',
        'uk': 'UK', 'united kingdom': 'UK',
        'canada': 'Canada', 'mexico': 'Mexico'
    }
    cleaned['country'] = cleaned['country'].str.lower().map(country_mapping)
    
    # 2. Clean abnormal age values
    cleaned['age'] = cleaned['age'].apply(
        lambda x: np.nan if (x < 0 or x > 120) else x
    )
    
    # 3. Remove near-duplicate names (normalize whitespace)
    cleaned['name'] = cleaned['name'].str.strip()
    cleaned = cleaned.drop_duplicates(subset=['name'], keep='first')
    
    return cleaned

# Apply the cleaning pipeline
final_cleaned_data = clean_dataset(dirty_data)

print("Before cleaning:")
print(f"  Rows: {len(dirty_data)}")
print(f"  Unique countries: {dirty_data['country'].nunique()}")
print(f"  Invalid ages: {((dirty_data['age'] < 0) | (dirty_data['age'] > 120)).sum()}")

print("\nAfter cleaning:")
print(f"  Rows: {len(final_cleaned_data)}")
print(f"  Unique countries: {final_cleaned_data['country'].nunique()}")
print(f"  Invalid ages: {((final_cleaned_data['age'] < 0) | (final_cleaned_data['age'] > 120)).sum()}")

print("\nCleaned dataset:")
print(final_cleaned_data[['customer_id', 'name', 'age', 'country', 'purchase_amount']])

### 🎯 Вправа на виклик

Тепер ваша черга! Нижче наведено новий рядок даних із кількома проблемами якості. Чи можете ви:

1. Визначити всі проблеми в цьому рядку
2. Написати код для виправлення кожної проблеми
3. Додати очищений рядок до набору даних

Ось проблемні дані:


In [ ]:
# New problematic row
new_row = pd.DataFrame({
    'customer_id': [13],
    'name': ['  Diana  Prince  '],  # Extra whitespace
    'age': [250],  # Impossible age
    'country': ['U.S.A.'],  # Inconsistent format
    'purchase_amount': [150.00]
})

print("New row to clean:")
print(new_row)

# TODO: Your code here to clean this row
# Hints:
# 1. Strip whitespace from the name
# 2. Check if the name is a duplicate (Diana Prince already exists)
# 3. Handle the impossible age value
# 4. Standardize the country name

# Example solution (uncomment and modify as needed):
# new_row_cleaned = new_row.copy()
# new_row_cleaned['name'] = new_row_cleaned['name'].str.strip()
# new_row_cleaned['age'] = np.nan  # Invalid age
# new_row_cleaned['country'] = 'USA'  # Standardized
# print("\nCleaned row:")
# print(new_row_cleaned)

### Основні моменти

1. **Непослідовні категорії** часто зустрічаються в реальних даних. Завжди перевіряйте унікальні значення та стандартизуйте їх за допомогою відповідностей або нечіткого зіставлення.

2. **Викиди** можуть суттєво вплинути на ваш аналіз. Використовуйте знання предметної області разом зі статистичними методами (IQR, Z-score) для їх виявлення.

3. **Майже дублікати** важче виявити, ніж точні дублікати. Розгляньте використання нечіткого зіставлення та нормалізації даних (перетворення в нижній регістр, видалення пробілів) для їх ідентифікації.

4. **Очищення даних — це ітеративний процес**. Можливо, вам доведеться застосувати кілька методів і переглянути результати перед тим, як завершити очищення набору даних.

5. **Документуйте свої рішення**. Фіксуйте, які кроки очищення ви застосували і чому, оскільки це важливо для відтворюваності та прозорості.

> **Найкраща практика:** Завжди зберігайте копію ваших оригінальних "брудних" даних. Ніколи не перезаписуйте вихідні файли даних — створюйте очищені версії з чіткими правилами іменування, наприклад, `data_cleaned.csv`.



---

**Відмова від відповідальності**:  
Цей документ було перекладено за допомогою сервісу автоматичного перекладу [Co-op Translator](https://github.com/Azure/co-op-translator). Хоча ми прагнемо до точності, звертаємо вашу увагу, що автоматичні переклади можуть містити помилки або неточності. Оригінальний документ мовою оригіналу слід вважати авторитетним джерелом. Для критично важливої інформації рекомендується професійний людський переклад. Ми не несемо відповідальності за будь-які непорозуміння або неправильні тлумачення, що виникли внаслідок використання цього перекладу.
